code/PaymentGatewayController.php

```diff
@@ -88,40 +99,149 @@ public function index()
+     * @param string $status The string identifier of the ServiceResponse
+     * @return void
+     */
+    protected function getServiceResponse($service, $status)
+    {
+        switch ($status) {
             case "complete":
-                $serviceResponse = $service->complete();
-                $response = $serviceResponse->redirectOrRespond();
+                $return = $service->complete();
                 break;
             case "notify":
-                $serviceResponse = $service->complete(array(), true);
-                $response = $serviceResponse->redirectOrRespond();
+                $return = $service->complete(array(), true);
                 break;
             case "cancel":
-                $serviceResponse = $service->cancel();
-                $response = $serviceResponse->redirectOrRespond();
+                $return = $service->cancel();
                 break;
             default:
-                $this->httpError(404, _t('Payment.InvalidUrl', 'Invalid payment url.'));
+                $return = null;
+        }
+
+        return $return;
+    }
+
+    /**
+     * The main action for handling all requests.
+     * It will redirect back to the application in all cases,
+     * but will not update the Payment/Transaction models if they are not found,
+     * or allowed to be updated.
+     */
+    public function index()
+    {
+        $response = null;
+        $payment = $this->getPaymentFromIdentifier($this->request->param('Identifier'));
+
+        if (!$payment) {
+            return $this->httpError(404, _t('Payment.NotFound', 'Payment could not be found.'));
+        }
+
+        $intent = $this->getPaymentIntent($payment->Status);
+
+        if (!$intent) {
+            return $this->httpError(403, _t('Payment.InvalidStatus', 'Invalid/unhandled payment status'));
+        }
+
+        $service = ServiceFactory::create()->getService($payment, $intent);
+        $service_response = $this->getServiceResponse($service, $this->request->param('Status'));
+
+        if (!$service_response) {
+            return $this->httpError(404, _t('Payment.InvalidUrl', 'Invalid payment url.'));
+        }
+        
+        $response = $service_response->redirectOrRespond();
+        
+        return $response;
+    }
+
+    /**
+     * Action used for handling static gateway requests
+     * (use this if your payment gateway doesnt handle
+     * dynamic callback URLs) 
+     */
+    public function gateway()
+    {
+        $response = null;
+        $gateway = $this->request->param("Gateway");
+        $identifier = null;
+
+        // Does the selected gateway allow static routes
+        if (!GatewayInfo::getConfigSetting($gateway, 'use_static_route')) {
+            return $this->httpError(404, _t('Payment.InvalidUrl', 'Invalid payment url.'));
+        }
+
+        $identifier = $this->getIdentifierFromRequest($this->request,$gateway);
+
+        if (!$identifier) {
```

Comment:
I think we can remove this check and just fail if there's no payment (a few lines below). I'd also suggest to just get the payment from the request (instead of the identifier). See explanation at line 262.